server/re/src/msi_genquery2.cpp

```diff
@@ -0,0 +1,315 @@
+///
+///         if (*ec < 0) {
+///             if (END_OF_RESULTSET == *ec) {
+///                 break;
+///             }
+///
+///             failmsg(*ec, "Unexpected error while iterating over GenQuery2 resultset.");
+///         }
+///
+///         msi_genquery2_column(*handle, '0', *coll_name); # Copy the COLL_NAME into *coll_name.
+///         msi_genquery2_column(*handle, '1', *data_name); # Copy the DATA_NAME into *data_name.
+///         writeLine("stdout", "logical path => [*coll_name/*data_name]");
+///     }
+///
+///     # Free any resources used. This is handled for you when the agent is shut down as well.
+///     msi_genquery2_free(*handle);
+/// }
+/// \endcode
+///
+/// \since 4.3.2
+auto msi_genquery2_next_row(MsParam* _handle, RuleExecInfo* _rei) -> int
+{
+    log_msi::trace(__func__);
+
+    IRODS_MSI_REQUIRE_VALID_POINTER(_handle);
+    IRODS_MSI_REQUIRE_VALID_POINTER(_rei);
+
+    IRODS_MSI_REQUIRE_VALID_POINTER(_handle->type);
+
+    IRODS_MSI_REQUIRE_TYPE(_handle->type, STR_MS_T);
+
+    IRODS_MSI_REQUIRE_VALID_POINTER(_handle->inOutStruct);
+
+    try {
+        const auto ctx_handle_index = std::stoll(static_cast<char*>(_handle->inOutStruct));
+
+        if (ctx_handle_index < 0) {
+            log_msi::error("{}: Unknown context handle.", __func__);
+            return SYS_INVALID_INPUT_PARAM;
+        }
+
+        auto iter = gq2_context.find(ctx_handle_index);
+
+        if (std::end(gq2_context) == iter) {
+            log_msi::error("{}: Unknown context handle.", __func__);
+        }
+
+        auto& ctx = iter->second;
+
+        if (ctx.current_row < static_cast<std::int32_t>(ctx.rows.size()) - 1) {
+            ++ctx.current_row;
+            log_msi::trace("{}: Incremented row position [{} => {}]. Returning 0.",
+                           __func__,
+                           ctx.current_row - 1,
+                           ctx.current_row);
+            return 0;
+        }
+
+        log_msi::trace("{}: Skipping increment of row position [current_row=[{}]]. Returning END_OF_RESULTSET ({}).",
+                       __func__,
+                       ctx.current_row,
+                       END_OF_RESULTSET);
+
+        return END_OF_RESULTSET;
+    }
+    catch (const std::exception& e) {
+        log_msi::error("{}: {}", __func__, e.what());
+        return SYS_LIBRARY_ERROR;
+    }
+} // msi_genquery2_next_row
+
+/// Reads the value of a column from a row within a GenQuery2 resultset.
+///
+/// \warning This microservice is experimental and may change in the future.
+/// \warning This microservice is NOT thread-safe.
+///
+/// GenQuery2 handles are only available to the agent which produced them.
+///
+/// \param[in]     _handle       The GenQuery2 handle.
+/// \param[in]     _column_index The index of the column to read. The index must be passed as a string.
+/// \param[in,out] _column_value The variable to write the value of the column to.
```

Comment:
Why `in,out` and not just `out`?

Reply:
I think `out` is the right answer.

`in,out` indicates a value of importance is passed into the function. `out` indicates the value held by the variable is not important and will be overridden under the right conditions.

Will update all occurrences of this.

The documentation for existing APIs should be updated to follow this as well.

Reply:
Done